prov/opx/src/fi_opx_init.c

```diff
@@ -0,0 +1,481 @@
+
+	if ((info->tx_attr) && ((info->tx_attr->caps | info->caps) != info->caps)) {
+FI_DBG_TRACE(fi_opx_global.prov, FI_LOG_EP_DATA, "info->tx_attr->caps = 0x%016lx, info->caps = 0x%016lx\n", info->tx_attr->caps, info->caps);
+		FI_LOG(fi_opx_global.prov, FI_LOG_DEBUG, FI_LOG_FABRIC,
+				"The tx_attr capabilities (0x%016lx) must be a subset of those requested of the associated endpoint (0x%016lx)",
+				info->tx_attr->caps, info->caps);
+		goto err;
+	}
+
+	if ((info->rx_attr) && ((info->rx_attr->caps | info->caps) != info->caps)) {
+FI_DBG_TRACE(fi_opx_global.prov, FI_LOG_EP_DATA, "info->rx_attr->caps = 0x%016lx, info->caps = 0x%016lx, (info->rx_attr->caps | info->caps) = 0x%016lx, ((info->rx_attr->caps | info->caps) ^ info->caps) = 0x%016lx\n", info->rx_attr->caps, info->caps, (info->rx_attr->caps | info->caps), ((info->rx_attr->caps | info->caps) ^ info->caps));
+		FI_LOG(fi_opx_global.prov, FI_LOG_DEBUG, FI_LOG_FABRIC,
+				"The rx_attr capabilities (0x%016lx) must be a subset of those requested of the associated endpoint (0x%016lx)",
+				info->rx_attr->caps, info->caps);
+		goto err;
+	}
+
+	switch (info->addr_format) {
+	case FI_ADDR_OPX:
+	case FI_FORMAT_UNSPEC:
+		break;
+	default:
+		FI_LOG(fi_opx_global.prov, FI_LOG_DEBUG, FI_LOG_FABRIC,
+				"unavailable [bad info->addr_format (%u)]",
+				info->addr_format);
+		goto err;
+	}
+
+	if (info->tx_attr) {
+		ret = fi_opx_check_tx_attr(info->tx_attr);
+		if (ret)
+			return ret;
+	}
+
+	if (info->rx_attr) {
+		ret = fi_opx_check_rx_attr(info->rx_attr);
+		if (ret)
+			return ret;
+	}
+
+	if (info->ep_attr) {
+		ret = fi_opx_check_ep_attr(info->ep_attr);
+		if (ret)
+			return ret;
+	}
+
+	if (info->domain_attr) {
+		ret = fi_opx_check_domain_attr(info->domain_attr);
+		if (ret)
+			return ret;
+	}
+
+	if (info->fabric_attr) {
+		ret = fi_opx_check_fabric_attr(info->fabric_attr);
+		if (ret)
+			return ret;
+	}
+
+	return 0;
+
+err:
+
+	errno = FI_ENODATA;
+	return -errno;
+}
+
+static int fi_opx_fillinfo(struct fi_info *fi, const char *node,
+		const char* service, const struct fi_info *hints,
+	        uint64_t flags)
+{
+	int ret;
+	uint64_t caps;
+	union fi_opx_addr *addr, *src_addr, *dest_addr;
+	uint32_t fmt;
+	size_t len;	
+
+	if (!fi)
+		goto err;
+
+	if (!hints && !node && !service)
+		goto err;
+
+	if (hints && (((hints->mode & FI_CONTEXT) != 0) && ((hints->mode & FI_CONTEXT2) == 0))) {
+		FI_WARN(fi_opx_global.prov, FI_LOG_FABRIC,
+			"FI_CONTEXT mode is not supported. Use FI_CONTEXT2 mode instead.\n");
+		errno = FI_ENODATA;
+		return -errno;
+	}
+
+	fi->next = NULL;
+	fi->caps = FI_OPX_DEFAULT_CAPS;
+
+	/* set the mode that we require */
+	fi->mode = FI_ASYNC_IOV;
+	fi->mode |= (FI_CONTEXT2);
+
+	/* clear modes that we do not require */
+	fi->mode &= (~FI_LOCAL_MR);
+	fi->mode &= (~FI_MSG_PREFIX);
+	fi->mode &= (~FI_CONTEXT);
+
+	fi->addr_format = FI_ADDR_OPX;
+	fi->src_addrlen = sizeof(union fi_opx_addr);
+	fi->dest_addrlen = sizeof(union fi_opx_addr);
```

Comment:
The dest_addr may be NULL if only a source address is specified.  May want to set this only when assigning a dest_addr.

Reply:
I will address this in the next push.